routing/payment_lifecycle.go

```diff
@@ -45,92 +63,156 @@ func (p *paymentLifecycle) resumePayment() ([32]byte, *route.Route, error) {
 			continue
 		}
 
-		_, err := shardHandler.collectResult(&a.HTLCAttemptInfo)
-		if err != nil {
-			return [32]byte{}, nil, err
-		}
+		shardHandler.collectResultAsync(
+			&a.HTLCAttemptInfo, shardOutcomes,
+		)
 	}
 
 	// We'll continue until either our payment succeeds, or we encounter a
 	// critical error during path finding.
 	for {
+
+		// Start by quickly checking if there are any outcomes already
+		// available to handle before we reevaluate our state.
+		for {
+			select {
+			case err := <-shardOutcomes:
+				if err != nil {
+					return [32]byte{}, nil, err
+				}
+				continue
+
+			case <-p.router.quit:
+				return [32]byte{}, nil, ErrRouterShuttingDown
+
+			default:
+			}
+
+			break
+		}
+
+		// Get all attempts made for this payment, and use that to
+		// deterimine the current status of the overall payment.
 		payment, err := p.router.cfg.Control.FetchPayment(
 			p.paymentHash,
 		)
 		if err != nil {
 			return [32]byte{}, nil, err
 		}
 
-		var attempt *channeldb.HTLCAttemptInfo
-		for _, a := range payment.HTLCs {
-			a := a
+		var (
+			// Keep track of whether we have any shards in-flight...
+			activeShards bool
 
-			// Terminal state, return the preimage and the route
-			// taken.
-			if a.Settle != nil {
-				return a.Settle.Preimage, &a.Route, nil
-			}
+			// ...and the total value of all in-flight and settled
+			// shards.
+			remValue = p.totalAmount
 
+			// If the paymennt is already marked failed, we'll get
+			// ready to terminate the moment all shards have
+			// returned.
+			terminate = payment.FailureReason != nil
+		)
+
+		for _, a := range payment.HTLCs {
+			// Ignore failed attempts.
 			if a.Failure != nil {
 				continue
 			}
 
-			attempt = &a.HTLCAttemptInfo
-			break
-		}
+			// Subtract settled or in flight attempts from the
+			// remaining value.
+			remValue -= a.Route.Amt()
```

Comment:
panic on underflow?

Reply:
Yeah, maybe we should handle this better. It can actually happen atm, since we don't validate the routes created. Will add validation at DB insertion.

Isn't a panic a bit overkill though?

Reply:
Did this: 9cd712105

lmk what you think.